### PR TITLE
Ghost Roles

### PR DESCRIPTION
**Pull request by Arrow768**
Makes the various ghost roles available via one spawn menu.

https://forums.aurorastation.org/topic/12346-ghost-roles/

**ToDo**

Addition of the following features:
- [x] Human Spawning
- [x] Name Changes
- [x] Appearance Changes
- [x] Logging when someone joins as a ghost role.
- [x] Unit Test Ghost Spawners

Integration of the following roles:
- [x] Mice
- [x] Maint Drones
- [x] CCIA Agents
- [x] CCIA Protection Detail
- [x] ERT Commander
- [x] TCFL Commander
- [x] Odin Checkpoint Guards
- [x] Merchant Assistant
- [x] Survivor Pod

Investigate the following features:
- [x] Search/Filter options for the GUI
- [x] List the number of available slots
- [x] Ghost Roles Created by Random Events (Escape Pod on the Asteroid, Visitor, ...)
- [ ] ~Defining Ghost Spawners via mapped in Objects~
- [ ] ~Allow people that can edit a role to edit the number of slots this role has~
- [ ] ~Listing them on the Job Menu?~

### CODE REVIEW

**Comment by alexpkeaton** [comment]
I played on a server that used a ghost role spawn menu (now defunct Oracle Station).  It was pretty sleek though this system is more all inclusive (Oracle's was basically a lavaland spawner - ERT/mice/drone sold separately).  I am not a code monkey but appreciate the concept.

Interested in seeing: 
-small chance of a survivor pod appearing somewhere in the mining asteroid.
-adminspawn zone stuff that can be turned on by staff (on other servers this might include a test/sandbox area or thunderdome)
-"need an antag" type roles, for example adminspawn event volunteers or the ability for admins to add a merc slot for a ghosted/empty slot

File: code/controllers/subsystems/ghostroles.dm [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	src.spawnpoints = SSghostroles.spawnpoints
+	src.spawners = SSghostroles.spawners
+
+/datum/controller/subsystem/ghostroles/New()
+	NEW_SS_GLOBAL(SSghostroles)
+	for(var/spawner in subtypesof(/datum/ghostspawner))
+		var/datum/ghostspawner/G = new spawner
+		//Check if we hae name, short_name and desc set
+		if(!G.short_name || !G.name || !G.desc)
+			continue
```
Comment (skull132) [inline code/controllers/subsystems/ghostroles.dm]: I'd personally add a `crash_with("Bad ghost spawner.")` here. The idea is that these will get created during unit testing and emitting a runtime will cause the unit tests to fail.

Thus, we detect a bad ghost spawner right off the bat! At least, if all ghost spawners are meant to adhere to these requirements.

File: code/controllers/subsystems/ghostroles.dm [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	
+/datum/controller/subsystem/ghostroles/proc/remove_spawnpoints(var/obj/effect/ghostspawner/G)
+	spawnpoints[G.identifier].Remove(G)
+	return
+
+//Returns the turf where the spawnpoint is located and updates the spawner to be used
+/datum/controller/subsystem/ghostroles/proc/get_spawnpoint(var/identifier, var/use = TRUE)
+	if(!identifier) //If no identifier ist set, return false
+		return FALSE
+	if(!spawnpoints[identifier] || !length(spawnpoints[identifier])) //If we have no spawnpoints for that identifier, return false
```
Comment (skull132) [inline code/controllers/subsystems/ghostroles.dm]:
```suggestion
	if(!length(spawnpoints[identifier])) //If we have no spawnpoints for that identifier, return false
```

`length(null)` is valid and returns `0` or `FALSE`.

File: code/controllers/subsystems/ghostroles.dm [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		VUEUI_SET_CHECK(newdata["spawners"][G.short_name]["tags"], G.tags, ., newdata)
+
+/datum/controller/subsystem/ghostroles/Topic(href, href_list)
+	var/datum/vueui/ui = href_list["vueui"]
+	if(!istype(ui))
+		return
+	if(href_list["action"] == "spawn")
+		var/spawner = href_list["spawner"]
+		var/datum/ghostspawner/S = spawners[spawner]
+		if(!S)
```
Comment (skull132) [inline code/controllers/subsystems/ghostroles.dm]:
```suggestion
		if(!istype(S))
```

File: code/controllers/subsystems/ghostroles.dm [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		if(!M)
+			to_chat(usr, "Unable to spawn: spawn_mob failed. Report this on GitHub")
+			return
+		if(!S.post_spawn(M))
+			to_chat(usr, "Unable to spawn: post_spawn failed. Report this on GitHub")
+			return
+		SSvueui.check_uis_for_change(src) //Make sure to update all the UIs so the count is updated
+	if(href_list["action"] == "enable")
+		var/datum/ghostspawner/S = spawners[href_list["spawner"]]
+		if(!S)
```
Comment (skull132) [inline code/controllers/subsystems/ghostroles.dm]:
```suggestion
		if(!istype(S))
```

File: code/controllers/subsystems/ghostroles.dm [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			return
+		if(!S.can_edit(usr))
+			return
+		if(!S.enabled)
+			S.enable()
+			to_chat(usr, "Ghost spawner enabled: [S.name]")
+			SSvueui.check_uis_for_change(src) //Update all the UIs to update the status of the spawner
+	if(href_list["action"] == "disable")
+		var/datum/ghostspawner/S = spawners[href_list["spawner"]]
+		if(!S)
```
Comment (skull132) [inline code/controllers/subsystems/ghostroles.dm]:
```suggestion
		if(!istype(S))
```

File: code/game/objects/items.dm [hunk 6]
```diff
@@ -385,6 +385,14 @@ var/list/global/slot_flags_enumeration = list(
 					allow = 1
 			if(!allow)
 				return 0
+		if(slot_in_belt)
+			var/allow = 0
+			if(H.belt && istype(H.belt, /obj/item/weapon/storage/belt))
```
Comment (skull132) [inline code/game/objects/items.dm]:
```suggestion
			if(istype(H.belt, /obj/item/weapon/storage/belt))
```

File: code/modules/ghostroles/spawner/base.dm [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	var/mob_name = FALSE //The name of that mob; If null prompts for it
+	var/mob_name_prefix = null //The prefix that should be applied to the mob (i.e. CCIAA, Tpr., Cmdr.)
+	var/mob_name_suffix = null //The suffix that should be applied to the mob name
+	
+
+//Return a error message if the user CANT see the ghost spawner. Otherwise FALSE
+/datum/ghostspawner/proc/cant_see(mob/user) //If the user can see the spawner in the menu
+	if(req_perms) //Only those with the correct flags can see restricted roles
+		if(check_rights(req_perms, show_msg=FALSE, user=user))
+			return FALSE //Return early and dont perform whitelist checks if staff flags are met
```
Comment (skull132) [inline code/modules/ghostroles/spawner/base.dm]:
```suggestion
			return null //Return early and dont perform whitelist checks if staff flags are met
```
I would return `null` over `FALSE` here. Semantics a bit, but it's more consistent: if you get non-null, it's an error object. null means no error object thus no error.

File: code/modules/ghostroles/spawner/base.dm [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	if(req_species_whitelist && !is_alien_whitelisted(user, req_species_whitelist))
+		return "Missing Species Whitelist"
+
+	if(jobban_job && jobban_isbanned(user,jobban_job))
+		return "Job Banned"
+	
+	if(!enabled && !can_edit(user)) //If its not enabled and the user cant edit it, dont show it
+		return "Currently Disabled"
+
+	return FALSE
```
Comment (skull132) [inline code/modules/ghostroles/spawner/base.dm]:
```suggestion
	return null
```

File: code/modules/ghostroles/spawner/base.dm [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		return "There is an administrative lock on entering the game"
+	if(SSticker.mode && SSticker.mode.explosion_in_progress)
+		return "The station is currently exploding"
+	if(max_count && count > max_count)
+		return "No more slots are available"
+	//Check if a spawnpoint is available
+	var/T = select_spawnpoint(FALSE)
+	if(!T)
+		return "No spawnpoint available"
+	return FALSE
```
Comment (skull132) [inline code/modules/ghostroles/spawner/base.dm]:
```suggestion
	return null
```

same here.

File: code/controllers/subsystems/ghostroles.dm (hunk 1 above)
Comment (Arrow768) [reply code/controllers/subsystems/ghostroles.dm]: As discussed, that is not possible due to child objects used to organize the spawners.

File: code/modules/ghostroles/spawner/base.dm (hunk 7 above)
Comment (Arrow768) [reply code/modules/ghostroles/spawner/base.dm]: Return False is intentional here, so I can work with 0 / text in the ui.

File: code/modules/ghostroles/spawner/base.dm (hunk 8 above)
Comment (Arrow768) [reply code/modules/ghostroles/spawner/base.dm]: See above

File: code/modules/ghostroles/spawner/base.dm (hunk 9 above)
Comment (Arrow768) [reply code/modules/ghostroles/spawner/base.dm]: See above

File: code/controllers/subsystems/ghostroles.dm [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+				G.spawn_mob()
+			return get_turf(G)
+
+/datum/controller/subsystem/ghostroles/proc/vui_interact(mob/user)
+	var/datum/vueui/ui = SSvueui.get_open_ui(user,src)
+	if(!ui)
+		ui = new(user,src,"misc-ghostspawner",950,700,"Ghost Role Spawner", nstate = interactive_state)
+	ui.open()
+
+/datum/controller/subsystem/ghostroles/vueui_data_change(var/list/newdata, var/mob/user, var/datum/vueui/ui)
```
[the remark this answers is not in the document]
Comment (Arrow768) [reply code/controllers/subsystems/ghostroles.dm]: The documentation uses newdata in the ui example.
If data should be used instead, the documentation should be updated.

File: code/modules/ghostroles/spawner/base.dm [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+		return cant_see
+	if(!istype(user, /mob/abstract/observer))
+		return "You are not a ghost."
+	if(!enabled) //If the spawner id disabled, we cant spawn in
+		return "This spawner is not enabled."
+	if(respawn_flag && !user.MayRespawn(0,respawn_flag))
+		return "You can not respawn at this time."
+	if(!config.enter_allowed)
+		return "There is an administrative lock on entering the game."
+	if(SSticker.mode && SSticker.mode.explosion_in_progress)
```
Comment (skull132) [inline code/modules/ghostroles/spawner/base.dm]:
```suggestion
	if(SSticker.mode?.explosion_in_progress)
```

File: code/controllers/subsystems/ghostroles.dm [hunk 12]
```diff
@@ -105,15 +105,15 @@
 /datum/controller/subsystem/ghostroles/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
 	if(!data)
 		. = data = list("current_tag"="All")
-	LAZYINITLIST(data["spawners"])
+	data["spawners"] = list()
```
Comment (Karolis2011) [inline code/controllers/subsystems/ghostroles.dm]: Please do LAZYINITLIST, even if list isn't lazy, it just does what we need - if value is not set, set it to empty list. This breaks stuff as it ALWAYS sets list to empty list.

File: code/controllers/subsystems/ghostroles.dm [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+		if(P.is_available())
+			if(use)
+				P.set_spawned()
+			return get_turf(P)
+
+/datum/controller/subsystem/ghostroles/proc/vui_interact(mob/user,var/spawnpoint=null)
+	var/datum/vueui/ui = SSvueui.get_open_ui(user,src)
+	if(!ui)
+		ui = new(user,src,"misc-ghostspawner",950,700,"Ghost Role Spawner", nstate = interactive_state)
+		ui.data = vueui_data_change(list("spawnpoint"=spawnpoint,"current_tag"="All"),user,ui)
```
Comment (Karolis2011) [inline code/controllers/subsystems/ghostroles.dm]: you now can actually use new thing I added - `ui,metadata` for this. It migth be nicier and better way to do spawnpoint lockdown.

File: vueui/src/components/view/misc/ghostspawner.vue [hunk 14]
```diff
@@ -0,0 +1,76 @@
+<template>
+  <div>
+    <div class="tagselector" v-show="!spawnpoint">
+      <vui-button v-for="(amount,tag) in tags" :key="tag" v-on:click="current_tag = tag" v-bind:class="{ selected: current_tag == tag}">{{tag}} ({{amount}})</vui-button>
+    </div>
+    <hr v-show="!spawnpoint">
```
Comment (Karolis2011) [inline vueui/src/components/view/misc/ghostspawner.vue]: Why this 'seperator' showing depends if spawnpoint is set or it is not set?

File: vueui/src/components/view/misc/ghostspawner.vue [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    th, td {
+      text-align: center;
+    }
+    .taction {
+      width: 200px;
+    }
+  }
+  .tagselector {
+    display: flex;
```
Comment (Karolis2011) [inline vueui/src/components/view/misc/ghostspawner.vue]: Ne flexbox, pls, because IE.

File: vueui/src/components/view/misc/ghostspawner.vue [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  table {
+    width: 100%;
+    th {
+      font-weight: bold;
+    }
+    th, td {
+      text-align: center;
+    }
+    .taction {
+      width: 200px;
```
Comment (Karolis2011) [inline vueui/src/components/view/misc/ghostspawner.vue]: Please do avoid using absolute sizing variables. _This is due to unwritten, unannounced new policy_.
Better make it automagically be minimum size without warping: https://stackoverflow.com/a/26983473/2425747

File: code/modules/ghostroles/spawner/human/merchant.dm [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+	special_role = "Merchants Assistant"
+	respawn_flag = null
+
+	mob_name = null
+
+/datum/ghostspawner/human/merchantass/can_edit(mob/user)
+    . = ..()
+    var/is_merchant = FALSE
+
+    if(istype(user, /mob/living/carbon/human))
```
Comment (skull132) [inline code/modules/ghostroles/spawner/human/merchant.dm]:
```suggestion
    if(ishuman(user))
```

File: vueui/src/components/view/misc/ghostspawner.vue (hunk 14 above)
Comment (Arrow768) [reply vueui/src/components/view/misc/ghostspawner.vue]: So there isnt a random horizontal line at the top if the ui for a specific spawnpoint is shown.

File: vueui/src/components/view/misc/ghostspawner.vue (hunk 15 above)
Comment (Arrow768) [reply vueui/src/components/view/misc/ghostspawner.vue]: _Alright, there you go. Have a table_

Comment (Arrow768) [reply vueui/src/components/view/misc/ghostspawner.vue]: Microsoft only supports the latest IE version on their clients.
On Windows 7, 8.1 and 10 that is 11.
IE 11 supports flexbox.

File: code/controllers/subsystems/ghostroles.dm (hunk 13 above)
Comment (Arrow768) [reply code/controllers/subsystems/ghostroles.dm]: After discussion in discord, opted not to implement that.

File: vueui/src/components/view/misc/ghostspawner.vue [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+      }
+    }
+  }
+}
+</script>
+
+<style lang="scss" scoped>
+  table {
+    width: 100%;
+    th.action, td.action {
```
Comment (Karolis2011) [inline vueui/src/components/view/misc/ghostspawner.vue]: Replace
```scss
    th.action, td.action {
      width: 1%;
      white-space: nowrap;
    }
    th, td {
      text-align: center;
      width: auto;
    }
```
with
```scss
    th, td {
      text-align: center;
      width: auto;
      &.action {
        width: 1%;
        white-space: nowrap;
      }
    }
```

Comment (Arrow768) [inline vueui/src/components/view/misc/ghostspawner.vue]:
```suggestion
    th, td {
      text-align: center;
      width: auto;
      &.action {
        width: 1%;
        white-space: nowrap;
      }
    }
```

File: code/unit_tests/spawner_tests.dm [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+		/datum/ghostspawner/human/admin,
+		/datum/ghostspawner/simplemob
+		)
+	var/failed_checks = 0
+	var/checks = 0
+
+	for(var/spawner in subtypesof(/datum/ghostspawner))
+		checks++
+		var/datum/ghostspawner/G = new spawner
+		if(is_type_in_list(G,ignore_spawners))
```
Comment (skull132) [inline code/unit_tests/spawner_tests.dm]: this doesn't work. It executes `istype`, which will mean that all subtypes of the ignored_spawners list will get ignored. As opposed to the explicit types noted.

Either use a typecache, or refactor this to the following:
```suggestion
		if(instances_of_type_in_list(G,ignore_spawners, strict = TRUE))
```

Comment (skull132) [reply code/unit_tests/spawner_tests.dm]: You might also want to print out the number of spawners actually checked. And verify it.

**Pull request merged**